learn/what_is_meilisearch/comparison_to_alternatives.md

```diff
@@ -75,31 +75,31 @@ Can't find a client you'd like us to support? [Submit your idea or vote for it](
 |   | Meilisearch | Algolia | Typesense | Elasticsearch |
 |---|:---:|:----:|:---:|:---:|
 | Typo tolerant  | ✅ | ✅ | ✅ | 🔶 <br>Needs to be specified by fuzzy queries |
-| Orderable ranking rules | ✅ | ✅ | 🔶 <br>Tie-breaking order is limited by a unique scoring rule | ❌|
-| Custom rules | ✅ | ✅ | 🔶 <br>Limited to one default sorting rule | 🔶 <br>Function score query
+| Orderable ranking rules | ✅ | ✅ | 🔶 <br>Field priority can be changed, but ranking rules order cannot be changed. | ❌|
+| Custom rules | ✅ | ✅ | ✅ | 🔶 <br>Function score query
 | Query field weights | ✅ | ✅ | ✅ | ✅ |
 | Synonyms | ✅ | ✅ | ✅ | ✅ |
 | Stop words | ✅ | ✅ | ❌ | ✅ |
 | Automatic language detection | ✅ | ✅ | ❌ | ❌ |
-| All language supports | ✅ | ✅ | ❌ <br> Only space separated | ✅ |
+| All language supports | ✅ | ✅ | ✅ | ✅ |
 
 ##### Security
 
 |   | Meilisearch | Algolia | Typesense | Elasticsearch |
 |---|:---:|:----:|:---:|:---:|
 | API Key Management | ✅ | ✅ | ✅ | ✅ |
-| Tenant tokens & multi-tenant indexes | ✅ <br> [Multitenancy support](/learn/security/tenant_tokens.md) | ✅  <br> Hard filters are not configurable per index for an end-user tenant key | ✅ <br> Hard filters are not configurable per index for an end-user tenant key | ✅ <br> Role-based |
+| Tenant tokens & multi-tenant indexes | ✅ <br> [Multitenancy support](/learn/security/tenant_tokens.md) | ✅  <br> Hard filters are not configurable per index for an end-user tenant key | ✅ | ✅ <br> Role-based |
 
 ##### Search
 
 |   | Meilisearch | Algolia | Typesense | Elasticsearch |
 |---|:---:|:----:|:---:|:---:|
 | Placeholder search | ✅ | ✅ | ✅ | ✅ |
 | Multi-index search | [2023](https://github.com/meilisearch/product/issues/74) | ✅ | ✅ | ✅ |
-| Exact phrase search | ✅ | ✅ | ❌ | ✅ |
+| Exact phrase search | ✅ | ✅ | ✅ | ✅ |
 | Geo search |  ✅  | ✅ | ✅ | ✅ |
-| Sort by  |  ✅  | 🔶 <br> Limited to one `sort_by` rule per index. Indexes may have to be duplicated for each sort field and sort order | 🔶 <br> Does not support sort on string field | ✅ |
-| Filtering |  ✅ <br> Support complex filter queries with an SQL-like syntax.  | 🔶 <br> Does not support `OR` operation across multiple fields | 🔶 <br> Does not support `OR` operation across multiple fields | ✅ |
+| Sort by  |  ✅  | 🔶 <br> Limited to one `sort_by` rule per index. Indexes may have to be duplicated for each sort field and sort order | ✅ | ✅ |
+| Filtering |  ✅ <br> Support complex filter queries with an SQL-like syntax.  | 🔶 <br> Does not support `OR` operation across multiple fields | ✅ | ✅ |
```

Comment:
Addressed this in v0.24

Reply:
I also read that sometimes, Typesense can be slow to answer as it does sort at query time [there even are workaround to this](https://github.com/typesense/typesense/issues/575#issuecomment-1096591171). We will inform users about this.

> Since Typesense dynamically sorts the results at query time, for stop words like `the` which occurs in a lot of documents, a large number of documents need to be processed to fetch results.
> 
> In the future, we plan to introduce a way to fetch documents (especially these types of single word searches) in a pre-sorted fashion. For now, removing such stop words or pre-caching them with a long TTL is the work around.

Reply:
Sure, feel free to mention this, but could you add that this only applies if the number of results fetched for a particular query is in the 10s of millions (count) range. This is not an issue if the number of results fetched are smaller than that.

Reply:
That's unfortunately not true. [As an example](https://songs-search.typesense.org/?s%5Bquery%5D=%2A&s%5BrefinementList%5D%5Bprimary_artist_name%5D%5B0%5D=Various%20Artists&s%5Bmenu%5D%5Brelease_decade%5D=2010s), when we search for less than 2 million documents the response time is pretty high (4s).

<img width="1231" alt="Capture d’écran 2023-02-16 à 23 47 03" src="https://user-images.githubusercontent.com/3610253/219504558-f1a1d633-ab64-4d04-acb3-6aa8ec3f09d9.png">

Reply:
In your screenshot, it says the number of results returned are 16.9M... So what I said still holds true: 

> this only applies if the number of results fetched for a particular query is in the 10s of millions (count) range

In your example, sorting ~17M records on the fly is indeed not as fast as we'd like it to be. But for smaller datasets below say 5M-10M should still be fast...

Reply:
Hum... no, my screenshot shows 1,695,074 documents, so 1,6M. That is not near what your explaining in your previous comment. And my second example shows an even worst case where Typesense answers in more than 2 second to sort half a million documents (660k).

Reply:
I indeed think warning readers about this unfortunate slowdown is required.

Reply:
Ah hmm... My bad. I misread the number in your screenshot. 

So yeah fair enough. Feel free to mention that for wild card queries `q=*` (which I believe you call "Placeholder search"), sorting is slow since it's done on the fly at runtime.

Reply:
Hum... what I understand is that Typesense is slow whenever there is a lot of documents to sort. Therefore, it can happen on a placeholder search, a placeholder search with filters (☝️), or any other requests that contains common words i.e. stop words, URL prefixes, URL domain names.

So, I would rather prefer that we explain that Typesense have performance issues for when more than 700k documents must be sorted.

Reply:
> what I understand is that Typesense is slow whenever there is a lot of documents to sort

You're right. That's the general case.

So to be more precise, I would say "When >700K results match for a given search query, queries might take several seconds, due to run-time sorting".

Reply:
After refining a little bit more the results and processing time of Typesense I found out that the engine [takes more than 3 seconds to sort approximately 28,000 documents](https://songs-search.typesense.org/?s%5Bquery%5D=%2A&s%5BrefinementList%5D%5Bprimary_artist_name%5D%5B0%5D=Пётр%20Ильич%20Чайковский). I would rather prefer that you write something like: « When >25K results match for a given search query, queries might take several seconds, due to run-time sorting », please.

Reply:
I did a bit more digging and my suspicion is that it's actually not the sorting that's the main cause of the slowness for these examples you shared. 

For eg, here's one example where ~1M documents are retrieved and sorted in 949ms:
<img width="700" alt="Screen Shot 2023-02-18 at 4 09 24 PM" src="https://user-images.githubusercontent.com/458383/219901568-deb86e4e-2f82-4394-99f3-d2fe78e82ae1.png">

Here's another example where ~600K documents are retrieved and sorted in 478ms:
<img width="700" alt="Screen Shot 2023-02-18 at 4 24 51 PM" src="https://user-images.githubusercontent.com/458383/219902024-06df0a41-348e-413e-8474-197c0aac9005.png">

So I take back what I said [here](https://github.com/meilisearch/documentation/pull/2174#discussion_r1110171297) - my general statement there does not hold true based on my examples above. 

It's likely a combination of filtering, placeholder search and faceting that's causing majority of the slowdown in the examples you shared. 

In your latest example, 3s to sort 28K documents definitely sounds more like a bug or edge case we probably didn't account for, given the shape of this particular dataset and query. We're going to look deeper into why that particular query is slow.

(Thank you for helping uncover these cases btw!)

In the meantime, I'd still be comfortable if you want to say that placeholder search or 1-3 letter prefix searches with filtering and faceting enabled might take several seconds when results fetched are over say 700K... But based on the examples I shared above, I don't think we can make a general statement about runtime sorting being slow.

Reply:
> In your latest example, 3s to sort 28K documents definitely sounds more like a bug or edge case we probably didn't account for, given the shape of this particular dataset and query.

There is not a [single](https://songs-search.typesense.org/?s%5Bquery%5D=%2A&s%5BrefinementList%5D%5Bprimary_artist_name%5D%5B0%5D=Johannes%20Brahms) [example](https://songs-search.typesense.org/?s%5Bquery%5D=%2A&s%5BrefinementList%5D%5Bprimary_artist_name%5D%5B0%5D=Franz%20Schubert) [that](https://songs-search.typesense.org/?s%5Bquery%5D=%2A&s%5BrefinementList%5D%5Bprimary_artist_name%5D%5B0%5D=Pearl%20Jam) [shows](https://songs-search.typesense.org/?s%5Bquery%5D=%2A&s%5BrefinementList%5D%5Bprimary_artist_name%5D%5B0%5D=Fryderyk%20Chopin) [the](https://songs-search.typesense.org/?s%5Bquery%5D=%2A&s%5BrefinementList%5D%5Bprimary_artist_name%5D%5B0%5D=Bob%20Dylan) [issue](https://songs-search.typesense.org/?s%5Bquery%5D=%2A&s%5BrefinementList%5D%5Bprimary_artist_name%5D%5B0%5D=J.%20K.%20Rowling). I am not sure that it is a bug or edge case but more a design issue, here.

> In the meantime, I'd still be comfortable if you want to say that placeholder search or 1-3 letter prefix searches with filtering and faceting enabled might take several seconds when results fetched are over say 700K...

Can we agree that the bound is way lower than 700k but more below 25k?

Reply:
Found some time to dig into this in depth. 

Here's what's happening: the search time you see displayed in the UI is actually the SUM of multiple search queries, embedded inside a single multi_search request (all these searches are required to show the facet counts in the RefinementList widgets). 

If you open up the network tab and view the actual request/response to Typesense, you'll notice a single multi_search request, but there are multiple search queries inside that request.

For eg, here's a seemingly bad one: filtering by [`Pearl Jam`](https://songs-search.typesense.org/?s%5Bquery%5D=%2A&s%5BrefinementList%5D%5Bprimary_artist_name%5D%5B0%5D=Pearl%20Jam) says: 29,935 results found - Searched 31,931,260 songs in 3400ms.

Now if you look at the network request, inside the single `multi_search` request, there are two search queries generated:

<img width="661" alt="Screen Shot 2023-02-20 at 12 31 26 AM" src="https://user-images.githubusercontent.com/458383/220030130-6c3233b6-f68d-40cb-ac5c-d52f1475998e.png">

Here's the corresponding responses for each search query:

<img width="802" alt="Screen Shot 2023-02-20 at 12 32 40 AM" src="https://user-images.githubusercontent.com/458383/220030346-381f9aa0-6652-4099-ab43-a51fe7cae149.png">

**Notice how the results for the first query which returns ~30K results is actually just 16ms**. The query that's actually slow is the one that facets on all the fields, across all 32M records, which takes about 3.4s.

This is the common thread I found across all the examples you shared. The actual search request itself is fast, but then there are other facet_by queries that need to be made to populate the facet counts for all the refinement list widgets (since they all need to be recomputed with each filter), and these are the time-consuming requests. 

*Side note:* these refinement list widgets used in the demo come from Instantsearch.js, so you will see the same set of requests generated even when you use these with Meilisearch and your adapter for Instantsearch.js. I wouldn't be surprised if you also see a 1s "search time" if you sum up the processing time of all the searches required to show results from a 28M record dataset, using 5 refinementList widgets. 

---

Here's a separate issue you helped uncover with [filtering](https://songs-search.typesense.org/?s%5Bquery%5D=%2A&s%5BrefinementList%5D%5Bgenres%5D%5B0%5D=Rock&s%5BrefinementList%5D%5Bprimary_artist_name%5D%5B0%5D=Various%20Artists&s%5Bmenu%5D%5Brelease_decade%5D=1970s). 

That example you shared currently does this multi_search request, with 4 different individual searches:

```
curl -s 'https://qtg5aekc2iosjh93p.a1.typesense.net/multi_search?use_cache=true&x-typesense-api-key=8hLCPSQTYcBuK29zY5q6Xhin7ONxHy99' \
  --data-raw '{
    "searches":[
        {"query_by":"primary_artist_name,title,album_name","query_by_weights":"2,2,1","sort_by":"_text_match(buckets: 10):desc","highlight_full_fields":"primary_artist_name,title,album_name","collection":"s","q":"*","facet_by":"genres,primary_artist_name,release_group_types,country,release_decade","filter_by":"genres:=[`Rock`] && primary_artist_name:=[`Various Artists`] && release_decade:=[`1970s`]","max_facet_values":20,"page":1,"per_page":15},
        {"query_by":"primary_artist_name,title,album_name","query_by_weights":"2,2,1","sort_by":"_text_match(buckets: 10):desc","highlight_full_fields":"primary_artist_name,title,album_name","collection":"s","q":"*","facet_by":"genres","filter_by":"primary_artist_name:=[`Various Artists`] && release_decade:=[`1970s`]","max_facet_values":20,"page":1,"per_page":1},
        {"query_by":"primary_artist_name,title,album_name","query_by_weights":"2,2,1","sort_by":"_text_match(buckets: 10):desc","highlight_full_fields":"primary_artist_name,title,album_name","collection":"s","q":"*","facet_by":"primary_artist_name","filter_by":"genres:=[`Rock`] && release_decade:=[`1970s`]","max_facet_values":20,"page":1,"per_page":1},
        {"query_by":"primary_artist_name,title,album_name","query_by_weights":"2,2,1","sort_by":"_text_match(buckets: 10):desc","highlight_full_fields":"primary_artist_name,title,album_name","collection":"s","q":"*","facet_by":"release_decade","filter_by":"genres:=[`Rock`] && primary_artist_name:=[`Various Artists`]","max_facet_values":20,"page":1,"per_page":1}]}' \
  | jq '.results[].search_time_ms'
761
645
358
626
```

By default the UI does an exact filtering `:=` on all the fields. Now notice how the search times change if I change it to `:` filtering:

```
curl -s 'https://qtg5aekc2iosjh93p.a1.typesense.net/multi_search?use_cache=true&x-typesense-api-key=8hLCPSQTYcBuK29zY5q6Xhin7ONxHy99' \
  --data-raw '{
    "searches":[
        {"query_by":"primary_artist_name,title,album_name","query_by_weights":"2,2,1","sort_by":"_text_match(buckets: 10):desc","highlight_full_fields":"primary_artist_name,title,album_name","collection":"s","q":"*","facet_by":"genres,primary_artist_name,release_group_types,country,release_decade","filter_by":"genres:[`Rock`] && primary_artist_name:[`Various Artists`] && release_decade:[`1970s`]","max_facet_values":20,"page":1,"per_page":15},
        {"query_by":"primary_artist_name,title,album_name","query_by_weights":"2,2,1","sort_by":"_text_match(buckets: 10):desc","highlight_full_fields":"primary_artist_name,title,album_name","collection":"s","q":"*","facet_by":"genres","filter_by":"primary_artist_name:[`Various Artists`] && release_decade:[`1970s`]","max_facet_values":20,"page":1,"per_page":1},
        {"query_by":"primary_artist_name,title,album_name","query_by_weights":"2,2,1","sort_by":"_text_match(buckets: 10):desc","highlight_full_fields":"primary_artist_name,title,album_name","collection":"s","q":"*","facet_by":"primary_artist_name","filter_by":"genres:[`Rock`] && release_decade:[`1970s`]","max_facet_values":20,"page":1,"per_page":1},
        {"query_by":"primary_artist_name,title,album_name","query_by_weights":"2,2,1","sort_by":"_text_match(buckets: 10):desc","highlight_full_fields":"primary_artist_name,title,album_name","collection":"s","q":"*","facet_by":"release_decade","filter_by":"genres:[`Rock`] && primary_artist_name:[`Various Artists`]","max_facet_values":20,"page":1,"per_page":1}]}' \
    | jq '.results[].search_time_ms'
273
263
147
250
```

This different in performance is a bug, which we already have a fix for in mind. We'll be pushing this fix out in 0.24.1.rc in the next week or two. 

---

Long story short:

It is not sorting that's slow, but _faceting_ on multiple fields on large datasets. I'd be ok with you saying that faceting on multiple fields on 10s of millions of records can take a couple of seconds. While we're going to look into how to improve this, this is not a bug - so feel free to mention this.

Reply:
Thank you very much for the time to dive into this issue. I took a little bit of time on my side too by trying another demo of Typesense, this time with a lot fewer documents in it i.e. 2 million.

> It is not sorting that's slow, but faceting on multiple fields on large datasets. I'd be ok with you saying that faceting on multiple fields on 10s of millions of records can take a couple of seconds. While we're going to look into how to improve this, this is not a bug - so feel free to mention this.

I absolutely agree with what you are saying here. However, it looks like the bounds are not exactly what I would have defined, but, again, I would have decided on a lower one. In [my example below](https://recipe-search.typesense.org/?r%5Bquery%5D=%2A) I ran a placeholder search on 2 millions documents and the facet retrieval took 730ms. Therefore, I would prefer to say that Typesense takes several seconds to answer when is needs to fetch facets on less than 10 million documents.

<img width="1258" alt="Capture d’écran 2023-02-20 à 21 33 33" src="https://user-images.githubusercontent.com/3610253/220194426-9dcdea56-3dda-4e65-b3c5-2c5a405264a4.png">
<img width="534" alt="Capture d’écran 2023-02-20 à 21 33 27" src="https://user-images.githubusercontent.com/3610253/220194462-7e16aafc-34e4-416d-899f-0bfdbd04609a.png">

What do you think? Could you update the "Faceted Search" category for Typesense to add something like:
> Faceting on a single field with less than 10 million documents can take several seconds.

Reply:
If you want to get more precise, it's really the number of unique values in the faceted field that determines how fast/slow facets are computed. So even though there are only 2M docs in this recipes example, the ingredients field is an array field and the number of unique facet values in this field across all 2M docs is in the 10s of millions range. 

```
curl -s 'https://qtg5aekc2iosjh93p.a1.typesense.net/multi_search?x-typesense-api-key=8hLCPSQTYcBuK29zY5q6Xhin7ONxHy99' \
  --data-raw '{"searches":[{"query_by":"title","sort_by":"","highlight_full_fields":"title","collection":"r","q":"*","facet_by":"ingredient_names","filter_by":"","max_facet_values":100,"page":1,"per_page":15}]}' \
  --compressed | jq '.results[0].facet_counts[0].counts[].count' | jq -s 'add'
11390471
```

So I still stand by my 10s of millions statement. If you want to refine it, you can say faceting on over 10s of millions of unique field values can take a couple of seconds.

Reply:
Unfortunately, this is different from what I witnessed below. There only is less than 6 million facet values, and Typesense takes more than 2 seconds to process them.
```
curl -s 'https://qtg5aekc2iosjh93p.a1.typesense.net/multi_search?x-typesense-api-key=8hLCPSQTYcBuK29zY5q6Xhin7ONxHy99' --data-raw '{"searches":[{"query_by":"primary_artist_name,title,album_name","query_by_weights":"2,2,1","sort_by":"","highlight_full_fields":"primary_artist_name,title,album_name","collection":"s","q":"*","facet_by":"primary_artist_name","max_facet_values":20,"page":1,"per_page":1}]}' --compressed | jq '.results[0].facet_counts[0].counts[].count' | jq -s 'add'
5933022
```

Reply:
`max_facet_values` is set to `20` in your query. So it's only returning the top 20 facets values in the API response, so you're counting a subset of the data. For eg, I you update `max_facet_values: 20000`, you'll still see ~2s response time, and the count is in the 10s of millions:

```
curl -s 'https://qtg5aekc2iosjh93p.a1.typesense.net/multi_search?x-typesense-api-key=8hLCPSQTYcBuK29zY5q6Xhin7ONxHy99' --data-raw '{"searches":[{"query_by":"primary_artist_name,title,album_name","query_by_weights":"2,2,1","sort_by":"","highlight_full_fields":"primary_artist_name,title,album_name","collection":"s","q":"*","facet_by":"primary_artist_name","max_facet_values":20000,"page":1,"per_page":1}]}' --compressed | jq '.results[0].facet_counts[0].counts[].count' | jq -s 'add'
19601674
```

Reply:
Thank you! So, what do you think of:
> Faceting can take several seconds when ~10 million facet values must be returned.

Reply:
> Thank you! So, what do you think of:
> 
> > Faceting can take several seconds when ~10 million facet values must be returned.

Yeah, I'd be ok with that.

Reply:
Can you update this PR with the sentence above added to Typesense in the « Faceted Search » category, please?

Reply:
Done.

Reply:
@Kerollmops May I know if Meilisearch estimates facet counts by default or if you return an exhaustive count by default?

Algolia for eg estimates facet counts by taking every third record and extrapolating. Do you do something similar?

In Typesense we return exhaustive facet counts by default, at least as of v0.24 and have the ability to turn on facet count estimation in the upcoming v0.25…

Curious to hear about Meilisearch’s approach.

Reply:
> May I know if Meilisearch estimates facet counts by default or if you return an exhaustive count by default?

Meilisearch returns accurate facet counts based on the estimated number of results we computed for a given query.

> Algolia for eg estimates facet counts by taking every third record and extrapolating. Do you do something similar?

What do you mean by _third record_?

Reply:
Hey @jasonbosco, can we consider [this performance issue](https://github.com/typesense/typesense/issues/914#issuecomment-1452228304) and change the sentence accordingly?

> If atleast 100000 hits are found for the search query, we will use 10% of the hits to estimate facet counts.

This isn't efficient and doesn't speed up the queries. I am not sure that [the previous sentence we agreed on](https://github.com/meilisearch/documentation/pull/2174/files#r1113828010) is accurate.

Reply:
> If atleast 100000 hits are found for the search query, we will use 10% of the hits to estimate facet counts.

This only happens when the user turns facet estimation on using `facet_sample_threshold` and `facet_sample_percent` parameters (added new in 0.25.0.rc). By default, we do not estimate facets or results, so my previous statement still holds true. 

On a side note, in that particular issue, I had run out of RAM which is what resulted in the slowness I saw. I can't replicate the issue anymore.

Reply:
Hey @jasonbosco,

>  I was able to just simulate direct curl commands similar to what the dynamic widgets would have made, and I was able to replicate the slowness. We'll look into this.

[There are performance issues on this subject](https://github.com/typesense/typesense/issues/914#issuecomment-1454236220). Could you add a line about that, please?

Just after you did your edit ☝️ I think we will be able to merge your PR. Thank you for making those changes. We are happy to describe Typesense better as it evolved all these past months.

Reply:
I can't add any other general statements to this PR, because this issue is isolated to this particular user's dataset and it looks like a bug.

For what it's worth, we handle 500M+ searches per month in Typesense Cloud, and I can confidently say that this is not a widespread issue for most users. 

So I'd appreciate not belaboring this topic and merging this PR, since the information that's currently public has factually incorrect statements about Typesense in other places.

Reply:
Let's merge that PR the way it is now and see if we need to make changes later.